src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -607,34 +607,82 @@ internal void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
 #nullable restore
             JsonSerializerOptions options);
 
-        internal virtual T ReadAsPropertyName(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        /// <summary>
+        /// Reads a value from a JSON property name.
+        /// </summary>
+        /// <param name="reader">The <see cref="Utf8JsonReader"/> to read from.</param>
+        /// <param name="typeToConvert">The <see cref="Type"/> being converted.</param>
+        /// <param name="options">The <see cref="JsonSerializerOptions"/> being used.</param>
+        /// <returns>The value that was converted.</returns>
+        public virtual T ReadAsPropertyName(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
             if (!IsInternalConverter && options.TryGetDefaultSimpleConverter(TypeToConvert, out JsonConverter? defaultConverter))
             {
                 // .NET 5 backward compatibility: hardcode the default converter for primitive key serialization.
                 Debug.Assert(defaultConverter.IsInternalConverter && defaultConverter is JsonConverter<T>);
-                return ((JsonConverter<T>)defaultConverter).ReadAsPropertyName(ref reader, TypeToConvert, options);
+                return ((JsonConverter<T>)defaultConverter).ReadAsPropertyNameCore(ref reader, TypeToConvert, options);
             }
 
             ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
             return default;
         }
 
-        internal virtual void WriteAsPropertyName(Utf8JsonWriter writer, T value, JsonSerializerOptions options, ref WriteStack state)
+        internal virtual T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            Debug.Assert(reader.TokenType == JsonTokenType.PropertyName);
+
+            long originalBytesConsumed = reader.BytesConsumed;
```

Comment:
I like this check, and we have similar checks elsewhere, however if they don't read the property, a JsonException will be thrown in the next Read* anyway.

Reply:
Note we could add full checks for JSON depth as well, like we do in OnTryRead\Write (which makes debugging so much easier) although implementations of this method should normally be simple values types and not as involved as Read\Write.

Reply:
> however if they don't read the property, a JsonException will be thrown in the next Read* anyway.

Making `Read*` calls should not be legal in this method, implementers are only expected to call `GetString()` or perhaps `ValueSpan`/`ValueSequence`.

Reply:
Don't the tests use an array as a key? (not just a simple Read*)

Reply:
Sure but if you take a look at the converter implementation it's simply calling `JsonSerializer.Deserialize<T>(reader.GetString())`.